### PR TITLE
feat(docker-codebuild): factory helper for using authentication and DockerHub images in CodeBuild

### LINKED ISSUE #43: Docker image Retrieval Helper
With the rate limiting now in place at DockerHub for retrieving images via unauthenticated pulls, we have been adding in authentication to CodeBuild/CodePipeline resources as appropriate.

In [dec-blueprints](https://github.com/ndlib/dec-blueprints/blob/main/src/dockerhub-image.ts), a helper class was created to assist with the creations of these CodeBuild build images. We haven't used the helper class in other places, but this is a regularly seen pattern of pulling a specific image as the CodeBuild image and therefore should probably have a reusable helper within ndlib-cdk.

### PR DESCRIPTION
**Pull request by ialford**
This method provides a simpler way to create a CodeBuild project that uses authenticated pulls from DockerHub. It does not affect the base CDK constructs, but rather allows the CodeBuilds to be made by only passing in two strings to the method.

Closes #43

### CODE REVIEW

File: src/docker-codebuild-action.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import { IBuildImage, LinuxBuildImage, PipelineProject, WindowsBuildImage } from '@aws-cdk/aws-codebuild';
+import { Secret } from '@aws-cdk/aws-secretsmanager';
+import cdk = require('@aws-cdk/core');
+
+export class DockerCodeBuildAction extends PipelineProject {
```
Comment (mlkirkpatrick) [inline src/docker-codebuild-action.ts:5]: I don't think you want to extend PipelineProject. It's most likely used in that context, but you don't need the PipelineProject behavior; these functions can probably just be a static class.

**Review by mlkirkpatrick — CHANGES_REQUESTED** [review]

**Review by rdoughty — APPROVED** [review]

**Comment by mlkirkpatrick** [comment]
Could you update the `fromWindowsDockerImage` to use `IDockerCodeBuildAction` as well?

**Review by SteveMattison — APPROVED** [review]

**Comment by ialford** [comment]
> Could you update the `fromWindowsDockerImage` to use `IDockerCodeBuildAction` as well?

bitmoji

Done.

File: README.md [hunk 2]
```diff
@@ -200,3 +200,55 @@ import { ArtifactBucket } from '@ndlib/ndlib-cdk';
...
+
+const stack = new cdk.Stack();
+const project = new PipelineProject(stack, `test-project`, {
+  environment: {
+    buildImage: DockerCodeBuildAction.fromLinuxDockerImage(
+      stack,
+      'alpine-build-image',
+      image: 'alpine:3',
+      credentialsContextKeyName: '/test/credentials',
+    ),
```
Comment (mlkirkpatrick) [inline README.md:223]: Update this

File: README.md [hunk 3]
```diff
@@ -200,3 +200,55 @@ import { ArtifactBucket } from '@ndlib/ndlib-cdk';
...
+    },
+    version: '0.2',
+  }),
+  environment: {
+    buildImage: DockerCodeBuildAction.fromWindowsDockerImage(
+      stack,
+      'iis-build-image',
+      image: 'mcr.microsoft.com/windows/servercore/iis',
+      credentialsContextKeyName: '/test/credentials',
+    ),
```
Comment (mlkirkpatrick) [inline README.md:251]: Update this

**Review by mlkirkpatrick — COMMENTED** [review]
Just some updates to the README then looks good!

File: README.md (hunk 2 above)
Comment (ialford) [reply README.md:223]: These have been updated from the previous version where it was just using a straight string. Was there something more that needed to be changed that I am missing?

Comment (mlkirkpatrick) [reply README.md:223]: No, that's all. I was just referring to the proper syntax for the new method signature. :)

Comment (ialford) [reply README.md:223]: Yeah, this is the updated syntax, matching what is used in the unit tests. If it's not what was expected or is unclear, let me know what else needs to be done.

Comment (mlkirkpatrick) [reply README.md:223]: It's missing the braces to make the 3rd parameter an actual object.

Comment (ialford) [reply README.md]: Good catch, have updated.

File: README.md (hunk 3 above)
Comment (ialford) [reply README.md]: Good catch, have updated.

**Review by mlkirkpatrick — APPROVED** [review]

**Pull request merged**